files/synchronize.py

```diff
@@ -345,15 +352,22 @@ def main():
     ssh_opts = '-S none'
 
     if not verify_host:
-      ssh_opts = '%s -o StrictHostKeyChecking=no' % ssh_opts
+        ssh_opts = '%s -o StrictHostKeyChecking=no' % ssh_opts
 
     if ssh_args:
-      ssh_opts = '%s %s' % (ssh_opts, ssh_args)
+        ssh_opts = '%s %s' % (ssh_opts, ssh_args)
 
     if dest_port != 22:
-        cmd += " --rsh 'ssh %s %s -o Port=%s'" % (private_key, ssh_opts, dest_port)
-    else:
-        cmd += " --rsh 'ssh %s %s'" % (private_key, ssh_opts)  # need ssh param
+        ssh_opts = '%s -o Port=%d' % (ssh_opts, dest_port)
+
+    if private_key:
+        cmd += " --rsh 'ssh %s %s'" % (private_key, ssh_opts)
+    elif ssh_passwd:
+        import os
+        sshpass_pipe = os.pipe()
+        os.write(sshpass_pipe[1], "{0}\n".format(ssh_passwd))
+        map(os.close, sshpass_pipe)
+        cmd += " --rsh 'sshpass -d %d ssh -l ass %s'" % (sshpass_pipe[0], ssh_opts)
```

Comment:
This does not work on my system

    hosts: 192.168.56.102
    staging_server: 192.168.56.101
    - synchronize: src=/etc
                   dest=/tmp
                   login_password=true
      delegate_to: "{{ staging_server }}"
  
From ps

    $ ps aux|grep sshpass
       sshpass -d 4 ssh -S none -o StrictHostKeyChecking=no -l root 192.168.56.102 rsync --server -logDtprze.iLs --log-format=%i --delay-updates . /tmp
    $  file /proc/pid_of_sshpass/fd
    0: broken symbolic link to `socket:[13047]'
    1: broken symbolic link to `socket:[13050]'
    2: broken symbolic link to `pipe:[13045]'
    3: symbolic link to `/dev/ptmx'
    4: symbolic link to `/dev/tty'
    5: symbolic link to `/dev/pts/1'

--

The most reliable way to fix this would be to use env variables, or to fix run_command function to support passing fd for input.

    $ vim   lib/ansible/module_utils/basic.py
    1700         if data:
    1701             st_in = subprocess.PIPE

This creates pipe for input to write data. This is usually <1>, but this can't be guaranteed. You could pass this fd no sshpass to read from.